server/routerlicious/packages/lambdas/src/alfred/index.ts

```diff
@@ -345,37 +375,6 @@ export function configureWebSocketServices(
 				);
 			}
 
```

Comment:
what about the below Promise.all where we join the socket to the rooms? is there a similar race there? probably smaller, but seems like the same problem could happen. Also, i don't necessarily expect it in this PR, but ideally the server can add some unit testing in this area to prevent regression. I would also say that most this code is pretty suspect, as its basically an unhandled promise as it is passed to a sync event handler. The whole flow probably need improvements to increase safety and testability.

Reply:
Yeah, in principle the same thing could happen. For local sever it probably can't right now because [the implementation of `socket.join()` in LocalWebSocket is actually synchronous](https://github.com/alexvy86/FluidFramework/blob/959f61b837084177829caa3baf0cd5525c496db8/server/routerlicious/packages/local-server/src/localWebSocketServer.ts#L29-L33). [Same for tinylicious' `SocketIoSocket`](https://socket.io/docs/v3/server-api/#socketjoinroom). Out in the wild, yeah, there's potentially still a race.

Reply:
@znewton @hedasilv maybe [that bit](https://github.com/alexvy86/FluidFramework/blob/959f61b837084177829caa3baf0cd5525c496db8/server/routerlicious/packages/lambdas/src/alfred/index.ts#L378-L396) could also be moved to before the clients list is computed? Right now it can happen that the channel subscriptions to the channel for a given client are set up, and then something else goes wrong that throws, so moving that to before getting the client list might be ok too?

Reply:
The important async logic that occurs when a client joins a session is:
- Get list of clients in the session (audience)
- Checking that the session is not full (max client count per session check)
- Adding the client to the list of clients in the session
- Inserting a ClientJoin op for that client in Kafka
- Subscribe the session/client to Redis pubsub & join the room (so that it gets ops/signals)
- Send connect_document_success

Note: those steps are not in the correct order and I'm assuming any user/token validation logic is occurring before all of this - so this PR is a good first step.

Resolving the race condition involves correctly ordering the get clients & redis subscribe call.
The redis pubsub subscribe / socket.io join must occur before the get clients call. That would ensure any ClientJoin/Leave signals are sent to this newly joining client even if we have yet to send them their connect_document_success message.

This order looks correct to me:
1. Checking that the session is not full (max client count per session check)
    1. use 1 redis command to count the number of client keys
1. Subscribe the session/client to Redis pubsub & join the room (so that it gets ops/signals)
1. Get list of clients in the session (audience)
1. Adding the client to the list of clients in the session
1. Inserting a ClientJoin op for that client in Kafka
1. Send connect_document_success

The tricky part here may be error handling since we'll have to ensure errors are handled correctly at any stage in the pipeline.

Reply:
I pushed the change so CI starts validating it too, just in case. I'll adjust if necessary.

Reply:
I was chatting with @alexvy86 and @znewton about this and even though I agree these changes are important and helpful, I was wondering if there could still be a chance for race condition?

For example:

1. ClientA starts connectDocument flow in Alfred instance 1. Starts the expensive auth and join room routines
2. ClientB starts connectDocument flow in Alfred instance 2. Also starts expensive auth and join room routines
3. Both of them complete those steps at the exact same time
4. Both read the client list from clientManager at the exact same time
5. Though this is more unlikely, they both would see an empty list, and the same symptom would happen

Is that possible?

Another potential issue: if at step 4 both clients read the client list at the same time, and the client list indicates we have room for only one more client (that is, the session is almost full), could we potentially add both clients to the session and thus lead to "disrespecting" the max connected client count?

I am wondering if there is any eventual consistency mechanism we could leverage to take care of those cases as well (though this would not help with the potential issue of overflowing max connected client count)

Reply:
Based on the discussions with @GaryWilber, @hedasilv, and @znewton , it seems like there's more work to be done in this area to tighten it up. I'll merge this PR as a small first step and mitigation for the behavior we already ran into.

Server folks, not sure where the best place would be to create tracking items in your backlogs to follow up on this; if you want to create them yourselves that's perfect, or if you point me to the right place I can do it too.